### DISCUSSION TITLE
DataGrid loses the sort when change ItemSource

### DISCUSSION THREAD

**Discussion by sn4k3** [AvaloniaUI/Avalonia#17746] [Q&A]
Currently have two `DataGrid`, the main DataGrid feeds the second DataGrid using it `SelectedItem.AnCollection`.
Works fine, however if user sorts the column the sort is lost after set a new `ItemSource` (Become the default unsorted grid). 
This is annoying because user have to re-sort every time he changes the selection in main grid.
Is this a bug or it's intended? 

I tried to use code to keep track of last sort, which I can but not of the sort direction...

**Comment by IanRawley** [comment]
How is the second DataGrid defined?  If the columns are auto-generated it's likely that changing the ItemsSource is causing the DataGrid to regenerate the columns, which would remove the sort as the appropriate column no longer exists.

**Reply by sn4k3** [reply]
Second DataGrid which lost the sort after ItemSource change:

```xml
<DataGrid Name="PingRepliesGrid"
          ItemsSource="{Binding #PingHostsGrid.((network:PingHost)SelectedItem).Replies}"
          MinHeight="200"
          CanUserResizeColumns="True"
          CanUserSortColumns="True"
          CanUserReorderColumns="True"
          ClipboardCopyMode="IncludeHeader"
          HeadersVisibility="All"
          IsReadOnly="True">
    <DataGrid.Styles>
        <Style Selector="DataGridRow">
            <Setter Property="Header" Value="{Binding $self.Index}"/>
        </Style>
    </DataGrid.Styles>
    <DataGrid.Columns>
        <DataGridTextColumn Header="Reply IP Address" Binding="{Binding IpAddressStr}" />
        <DataGridTextColumn Header="Time" Binding="{Binding Time}" />
        <DataGridTextColumn Header="TTL" Binding="{Binding Ttl}" />
        <DataGridTextColumn Header="Buffer size" Binding="{Binding BufferLength}" />
        <DataGridTextColumn Header="Sent on" Binding="{Binding SentOn}" />
        <DataGridTextColumn Header="Status" Binding="{Binding PingStatus}" />
        <DataGridTextColumn Header="Error message" Binding="{Binding ErrorMessage}" />
    </DataGrid.Columns>
</DataGrid>
```

The problem with code-behind to get around this problem is this property: `internal ListSortDirection? CurrentSortingState` this should be public instead. I can't understand why many useful properties in Avalonia are not public, since setter is already private...
